File: 22Adams_Bashforth/adams_bashforth3.py
import numpy as np
import sympy as sy
from sympy import symbols
import matplotlib.pyplot as plt



def adams_bashforth3(f,x_0,y_0,a,b,numeroPuntos):
    h = (b-a)/(numeroPuntos - 1)
    vectorX = obtenerVectorX(h,numeroPuntos,a,b)
    XY = euler(f,x_0,y_0,a,b,numeroPuntos)
    Y_1 = XY[1]
    Y_2 = XY[2]
    XY = adams_bashforth3_aux(f,vectorX,h,x_0,y_0,Y_1,Y_2)
    print('x',XY[0])
    print('y',XY[1])
    plt.plot(XY[0],XY[1])
    plt.show()
def adams_bashforth3_aux(f,vectorX,h,x_0,y_0,Y_1,Y_2):
    f = sy.sympify(f)
    x,y = symbols('x y')
    contador = 0
    m = len(vectorX)
    Y = np.zeros([m])
    Y[contador] = y_0
    Y[contador + 1] = Y_1
    Y[contador + 2] = Y_2
    x_n = vectorX[contador]
    y_n = Y_1
    funcion2 = f.evalf(subs={x:x_n,y:y_0})
    contador += 1
    x_n = vectorX[contador]
    y_n = Y[contador]
    funcion1 = f.evalf(subs={x:x_n,y:y_0})
    contador += 1
    x_n = vectorX[contador]
    y_n = Y[contador]
    funcion0 = f.evalf(subs={x:x_n,y:y_0})
    Y[contador + 1] = Y[contador] + (h/16)*(23*funcion0 - 16*funcion1 + 5*funcion2)
    contador += 1
    while contador != m:
        x_n = vectorX[contador - 1]
        y_n = Y[contador-1]
        funcion2 = f.evalf(subs={x:x_n,y:y_0})
        x_n = vectorX[contador - 2]
        y_n = Y[contador - 2]
        funcion1 = f.evalf(subs={x:x_n,y:y_0})
        x_n = vectorX[contador - 3]
        y_n = Y[contador - 3]
        funcion0 = f.evalf(subs={x:x_n,y:y_0})
        Y[contador] = Y[contador - 1] +(h/16)*(23*funcion2 - 16*funcion1+ 5*funcion0)
        contador += 1
    return [vectorX, Y]
    
    


def obtenerVectorX(h,numeroPuntos,a,b):
    vectorX = np.zeros([numeroPuntos])
    contador = 0

    while contador <= numeroPuntos-1:
        if contador == 0:
            vectorX[contador] = a
            contador += 1
        elif contador == numeroPuntos:
            vectorX[contador] = b
            contador += 1
        else:
            a = a + h
            vectorX[contador] = a
            contador += 1
    
    return vectorX


def euler (f,x_0,y_0,a,b,numeroPuntos):
    h = (b-a)/(numeroPuntos - 1)

    vectorX = obtenerVectorX(h,numeroPuntos,a,b)
    
    XY = vectoresXY(f,x_0,y_0,h,vectorX)
    return XY[1]
def obtenerVectorX(h,numeroPuntos,a,b):
    vectorX = np.zeros([numeroPuntos])
    contador = 0

    while contador <= numeroPuntos-1:
        if contador == 0:
            vectorX[contador] = a
            contador += 1
        elif contador == numeroPuntos:
            vectorX[contador] = b
            contador += 1
        else:
            a = a + h
            vectorX[contador] = a
            contador += 1
    
    return vectorX
    


def vectoresXY(f,x_0,y_0,h,vectorX):
    x,y = symbols('x y')
    contador = 0
    m = len(vectorX)
    Y = np.zeros([m])
    
    Y[contador] = y_0
    f = sy.sympify(f)
    contador += 1
    while contador != m:
        
        funcion = f.evalf(subs={x:vectorX[contador-1],y:Y[contador - 1]})

        Y[contador] = Y[contador - 1] + h*(funcion)

        contador += 1
    
    return [vectorX, Y]

adams_bashforth3('-x*y + 4*x/y',0,1,0,1,11)

File: 22Adams_Bashforth/adams_bashforth3.m
function resultado = adams_bashforth3(f,x_0,y_0,a,b,numeroPuntos)
  h = (b - a)/(numeroPuntos - 1);
  vectorX = obtenerVectorX(h,numeroPuntos,a,b);
  [X Y] = euler(f,x_0,y_0,a,b,numeroPuntos);
  Y_1 = Y(2);
  Y_2 = Y(3);
  [X Y] = adams_bashforth3_aux(f,vectorX,h,x_0,y_0,Y_1,Y_2);
  plot(X,Y)
endfunction
function [X Y] = adams_bashforth3_aux(f,vectorX,h,x_0,y_0,Y_1,Y_2)
  syms x y
  m = length(vectorX);
  Y = zeros(1,m);
  contador = 1;
  Y(contador) = y_0;
  Y(contador + 1) = Y_1;
  Y(contador + 2) = Y_2;
  x = vectorX(contador);
  y = Y_1;
  funcion2 = eval(f);
  contador = contador + 1;
  x = vectorX(contador);
  y = Y(contador);
  funcion1 = eval(f);
  contador = contador + 1;
  x = vectorX(contador);
  y = Y(contador);
  funcion0 = eval(f);
  Y(contador + 1) = Y(contador) + (h/16)*(23*funcion0 - 16*funcion1 + 5*funcion2);
  contador = contador + 1;
  while contador != m
    x = vectorX(contador - 1);
    y = Y(contador - 1);
    funcion2 = eval(f);
    x = vectorX(contador - 2);
    y = Y(contador - 2);
    funcion1 = eval(f);
    x = vectorX(contador - 3);
    y = Y(contador - 3);
    funcion0 = eval(f);
    Y(contador) = Y(contador - 1) + (h/16)*(23*funcion2 - 16*funcion1 + 5* funcion0);
    contador = contador + 1;
  endwhile
  X = vectorX;
  return
endfunction

function [X Y] = euler(f,x_0,y_0,a,b,numeroPuntos)
  pkg load symbolic
  h = (b-a)/(numeroPuntos - 1);
  
  vectorX = obtenerVectorX(h,numeroPuntos,a,b);
  
  [X Y] = vectoresXY(f,x_0,y_0,h,vectorX)
  return
endfunction

function vectorX = obtenerVectorX(h,numeroPuntos,a,b)
  vectorX = zeros(1,numeroPuntos);
  contador = 1;
  while contador <= numeroPuntos
    if contador == 1
      vectorX(contador) = a;
      contador = contador + 1;
    elseif contador == numeroPuntos;
      vectorX(contador) = b;
      contador = contador + 1;
    else
      a = a+h;
      vectorX(contador) = a;
      contador = contador + 1; 
    endif
  
  endwhile
  return 
endfunction


function [X Y] = vectoresXY(f,x_0,y_0,h,vectorX)
  syms x y;
  m = length(vectorX);
  contador = 1;
  
  Y = zeros(1,vectorX);
  Y(contador) = y_0;
  contador = 2;
  
  while contador != m + 1
    x = vectorX(contador - 1);
    y = Y(contador - 1);
    funcion = eval(f);
    Y(contador) = Y(contador-1) + h*(funcion);
    
    contador = contador + 1;
    
  endwhile
  X = vectorX;
  return
  
endfunction